lib/src/main/scala/spinal/lib/Stream.scala

```diff
@@ -555,29 +557,45 @@ class Stream[T <: Data](val payloadType :  HardType[T]) extends Bundle with IMas
     }
   }
 
+  def formalIsValid(payloadInvariance : Boolean = true) = new Composite(this, "formalIsValid") {
+    import spinal.core.formal._
+    val wasStall = past(isStall) init(False)
+    val checkValidHandshake = Mux(wasStall, valid, True)
+
+    val priorValidPayload = RegNextWhen(payload, valid)
+    val checkValidPayloadInvariance = Mux(wasStall && Bool(payloadInvariance),
+      priorValidPayload === payload,
+      True)
+
+    val isValid = checkValidHandshake && checkValidPayloadInvariance
+  }.isValid
+
+  override def formalIsProducerValid() = formalIsValid(true)
+
+  def formalAsserts(payloadInvariance : Boolean = true)(implicit loc : Location, useAssumes : Boolean = false) = new Composite(this, if(useAssumes) "assumes" else "asserts") {
+    import spinal.lib.formal._
+    import spinal.lib.formal.HasFormalAsserts._
+
+    val stack = ScalaLocated.long
+    when(past(isStall) init(False)) {
+      assertOrAssume(valid,  "Stream transaction disappeared:\n" + stack)
+      if(payloadInvariance) assertOrAssume( RegNextWhen(payload, valid) === payload, "Stream transaction payload changed:\n" + stack)
```

Comment:
according to previous disscusion the assume is not usable in hierachical verification, what does this assertOrAssume used for?
BTW, it seems that formalIsValid do what formalAsserts do?